core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -924,6 +934,7 @@ public List<ManifestFile> apply(TableMetadata base, Snapshot snapshot) {
                         != ManifestWriter
                             .UNASSIGNED_SEQ) // filter out unassigned in rewritten manifests
             .reduce(base.lastSequenceNumber(), Math::min);
+    long minDataSequenceNumber = Math.min(minNewFileSequenceNumber, minExistingDataSequenceNumber);
```

Comment:
The logic for calculating the min sequence number is getting too long to embed here. I think it should be moved to a separate private method:

```java
  private long minDataSequenceNumber() {
    long minAddedDataSequenceNumber = addedDataFiles().stream()
        .map(ContentFile::dataSequenceNumber)
        .filter(Objects::nonNull)
        .filter(seq -> seq >= 0)
        .reduce(base.nextSequenceNumber(), Math::min);

    long minExistingDataSequenceNumber =
        filtered.stream()
            .map(ManifestFile::minSequenceNumber)
            .filter(
                seq ->
                    seq
                        != ManifestWriter
                            .UNASSIGNED_SEQ) // filter out unassigned in rewritten manifests
            .reduce(base.lastSequenceNumber(), Math::min);
    long minDataSequenceNumber = Math.min(minAddedDataSequenceNumber, minExistingDataSequenceNumber)

    return Math.min(Math.min(minAddedDataSequenceNumber, minExistingDataSequenceNumber), newDataFilesDataSequenceNumber);
  }
```

This also fixes the awkwardness of deciding whether to use `base.nextSequenceNumber()` or `newDataFilesDataSequenceNumber`.

I also agree with @amogh-jahagirdar that checking `addedDataFiles()` is not currently necessary, but it seems like a good idea to future-proof this issue in case we change the code again. This bug was likely introduced when we added `newDataFilesDataSequenceNumber`.